### PR TITLE
Consolidate Kafka metrics in Hazelcast

### PR DESCRIPTION
**Pull request by teivah**
* Consolidation of the Kafka metrics in an Hazelcast map (embedded mode)
* Modification of the `/broker` endpoint to `/brokers/current`
* New endpoint `/brokers` to expose get the metrics for all the brokers

### CODE REVIEW

File: monitoring/weigh-bridge/src/main/java/com/hotels/road/weighbridge/WeighBridgeController.java [hunk 1]
```diff
@@ -15,22 +15,32 @@
...
 
-  @GetMapping("/broker")
-  public Broker get() {
+  @GetMapping("/brokers/current")
+  public Broker getCurrentBroker() {
     return brokerReference.get();
   }
+
+  @GetMapping("/brokers")
+  public List<Broker> get() {
```
Comment (nahguam) [inline monitoring/weigh-bridge/src/main/java/com/hotels/road/weighbridge/WeighBridgeController.java]: `getAllBrokers` to be consistent with the other method rename.

**Comment by nahguam** [comment]
please add a changelog entry

File: monitoring/weigh-bridge/src/main/resources/hazelcast.xml [hunk 2]
```diff
@@ -0,0 +1,46 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
[the remark this answers is not in the document]
Comment (nahguam) [reply monitoring/weigh-bridge/src/main/resources/hazelcast.xml]: Looks like you can create a `com.hazelcast.config.Config` `@Bean`

File: monitoring/weigh-bridge/src/main/resources/hazelcast.xml [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    <password>dev</password>
+  </group>
+  <management-center enabled="true">http://localhost:8081/mancenter-3.3.3</management-center>
+  <network>
+    <port auto-increment="true" port-count="100">5701</port>
+    <outbound-ports>
+      <ports>0</ports>
+    </outbound-ports>
+    <join>
+      <multicast enabled="true">
```
Comment (Noddy76) [inline monitoring/weigh-bridge/src/main/resources/hazelcast.xml]: I'm pretty sure that multicast will not work on the Kubernetes network.

For the session store implementation I've been looking at [hazelcast-kubernetes](https://github.com/hazelcast/hazelcast-kubernetes) to do the discovery via DNS on a headless service.

File: monitoring/weigh-bridge/src/test/java/com/hotels/road/weighbridge/WeighBridgeControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.junit.MockitoJUnitRunner;
+
+import com.sun.tools.javac.util.List;
```
[the remark this answers is not in the document]
Comment (teivah) [reply monitoring/weigh-bridge/src/test/java/com/hotels/road/weighbridge/WeighBridgeControllerTest.java]: Oh that's why I had this `List.of`:) Sorry

**Comment by teivah** [comment]
Missing:
- Hazelcast configuration (pending @Noddy76's work on https://github.com/HotelsDotCom/data-highway/pull/28)

File: monitoring/weigh-bridge/src/test/resources/hazelcast.xml [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    <password>dev</password>
+  </group>
+  <management-center enabled="true">http://localhost:8081/mancenter-3.3.3</management-center>
+  <network>
+    <port auto-increment="true" port-count="100">5701</port>
+    <outbound-ports>
+      <ports>0</ports>
+    </outbound-ports>
+    <join>
+      <multicast enabled="true">
```
Comment (Noddy76) [inline monitoring/weigh-bridge/src/test/resources/hazelcast.xml]: multicast shouldn't be used for tests as it has the possibility to bring in tests running concurrently on the same network.

File: monitoring/weigh-bridge/src/main/resources/hazelcast.xml (hunk 2 above)
[the remark this answers is not in the document]
Comment (teivah) [reply monitoring/weigh-bridge/src/main/resources/hazelcast.xml]: Done

File: monitoring/weigh-bridge/src/main/java/com/hotels/road/weighbridge/model/BrokerSerializer.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
+package com.hotels.road.weighbridge.model;
```
[the remark this answers is not in the document]
Comment (teivah) [reply monitoring/weigh-bridge/src/main/java/com/hotels/road/weighbridge/model/BrokerSerializer.java]: Done.

File: monitoring/weigh-bridge/src/main/resources/hazelcast.xml [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+        <multicast-group>224.2.2.3</multicast-group>
```
[the remark this answers is not in the document]
Comment (teivah) [reply monitoring/weigh-bridge/src/main/resources/hazelcast.xml]: This file is only used locally. The k8s Hazelcast plugin is configured in road-charts

File: monitoring/weigh-bridge/src/test/java/com/hotels/road/weighbridge/WeighBridgeControllerTest.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    final Broker broker1 = new Broker(1, "rack", new ArrayList<>());
+    final Broker broker2 = new Broker(2, "rack", new ArrayList<>());
+    final Broker broker3 = new Broker(3, "rack", new ArrayList<>());
+    final HashMap<Integer, Broker> map = new HashMap<>();
+    map.put(1, broker1);
+    map.put(2, broker2);
+    map.put(3, broker3);
+    underTest = new WeighBridgeController(new AtomicReference<>(broker1), map);
+
+    assertEquals(List.of(broker1, broker2, broker3), underTest.get());
```
[the remark this answers is not in the document]
Comment (teivah) [reply monitoring/weigh-bridge/src/test/java/com/hotels/road/weighbridge/WeighBridgeControllerTest.java]: Done

File: monitoring/weigh-bridge/src/main/resources/hazelcast.xml [hunk 9]
```diff
@@ -0,0 +1,46 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<hazelcast xsi:schemaLocation="http://www.hazelcast.com/schema/config hazelcast-config-3.3.xsd"
+           xmlns="http://www.hazelcast.com/schema/config"
+           xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
+  <group>
+    <name>dev</name>
+    <password>dev</password>
```
[the remark this answers is not in the document]
Comment (teivah) [reply monitoring/weigh-bridge/src/main/resources/hazelcast.xml]: Done on road-charts.

File: monitoring/weigh-bridge/src/main/resources/hazelcast.xml (hunk 2 above)
[the remark this answers is not in the document]
Comment (teivah) [reply monitoring/weigh-bridge/src/main/resources/hazelcast.xml]: See road-charts configuration

File: monitoring/weigh-bridge/src/main/java/com/hotels/road/weighbridge/model/BrokerSerializer.java [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+  }
+
+  @Override
+  public int getTypeId() {
+    return 1;
+  }
+
+  @Override
+  public void destroy() {
+
```
[the remark this answers is not in the document]
Comment (teivah) [reply monitoring/weigh-bridge/src/main/java/com/hotels/road/weighbridge/model/BrokerSerializer.java]: Fixed.

File: monitoring/weigh-bridge/src/main/java/com/hotels/road/weighbridge/model/BrokerSerializer.java [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.StreamSerializer;
+
+@Component
+public class BrokerSerializer implements StreamSerializer<Broker> {
+
+  @Override
+  public void write(ObjectDataOutput out, Broker broker) throws IOException {
+    ObjectMapper mapper = new ObjectMapper();
```
[the remark this answers is not in the document]
Comment (teivah) [reply monitoring/weigh-bridge/src/main/java/com/hotels/road/weighbridge/model/BrokerSerializer.java]: Fixed.

File: monitoring/weigh-bridge/src/main/java/com/hotels/road/weighbridge/WeighBridgeController.java (hunk 1 above)
Comment (teivah) [reply monitoring/weigh-bridge/src/main/java/com/hotels/road/weighbridge/WeighBridgeController.java]: Fixed.

File: monitoring/weigh-bridge/src/main/resources/hazelcast.xml (hunk 3 above)
Comment (teivah) [reply monitoring/weigh-bridge/src/main/resources/hazelcast.xml]: See road-charts configuration.

File: monitoring/weigh-bridge/pom.xml [hunk 12]
```diff
@@ -14,6 +14,7 @@
   <properties>
     <docker.name>road-weigh-bridge</docker.name>
     <docker.container.port>8081</docker.container.port>
+    <hazelcast-spring.version>3.11.1</hazelcast-spring.version>
```
[the remark this answers is not in the document]
Comment (teivah) [reply monitoring/weigh-bridge/pom.xml]: Done.

File: monitoring/weigh-bridge/src/main/java/com/hotels/road/weighbridge/model/BrokerSerializer.java [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.StreamSerializer;
+
+@Component
+public class BrokerSerializer implements StreamSerializer<Broker> {
+
+  private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
```
[the remark this answers is not in the document]
Comment (teivah) [reply monitoring/weigh-bridge/src/main/java/com/hotels/road/weighbridge/model/BrokerSerializer.java]: Done.

File: monitoring/weigh-bridge/src/test/resources/hazelcast.xml (hunk 5 above)
Comment (teivah) [reply monitoring/weigh-bridge/src/test/resources/hazelcast.xml]: Changed.

**Pull request merged**